sw/airborne/arch/chibios/mcu_periph/uart_arch.c

```diff
@@ -0,0 +1,681 @@
+/*
+ * Copyright (C) 2013 AggieAir, A Remote Sensing Unmanned Aerial System for Scientific Applications
+ * Utah State University, http://aggieair.usu.edu/
+ *
+ * Michal Podhradsky (michal.podhradsky@aggiemail.usu.edu)
+ * Calvin Coopmans (c.r.coopmans@ieee.org)
+ *
+ * Alexandre Bustico <alexandre.bustico@enac.fr>
+ * Gautier Hattenberger <gautier.hattenberger@enac.fr>
+ *
+ * This file is part of paparazzi.
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+/**
+ * @file arch/chibios/mcu_periph/uart_arch.c
+ * UART/Serial driver implementation for ChibiOS arch
+ *
+ * ChibiOS has a high level Serial Driver, for Paparazzi it is more convenient
+ * than pure UART driver (which needs callbacks etc.). This implementation is
+ * asynchronous and the RX thread has to use event flags. See ChibiOS documen-
+ * tation.
+ */
+#include "mcu_periph/uart_arch.h"
+#include <ch.h>
+#include <hal.h>
+
+struct SerialInit {
+  semaphore_t *rx_sem;
+  semaphore_t *tx_sem;
+  mutex_t *rx_mtx;
+  mutex_t *tx_mtx;
+};
+
+/**
+ * RX handler
+ */
+static void handle_uart_rx(struct uart_periph *p)
+{
+  // wait for next incoming byte
+  uint8_t c = sdGet((SerialDriver*)(p->reg_addr));
+
+  struct SerialInit *init_struct = (struct SerialInit*)(p->init_struct);
+  chMtxLock(init_struct->rx_mtx);
+  uint16_t temp = (p->rx_insert_idx + 1) % UART_RX_BUFFER_SIZE;;
+  // insert new byte
+  p->rx_buf[p->rx_insert_idx] = c;
+  // check for more room in queue
+  if (temp != p->rx_extract_idx) {
+    p->rx_insert_idx = temp;  // update insert index
+  }
+  chMtxUnlock(init_struct->rx_mtx);
+  chSemSignal(init_struct->rx_sem);
+}
+
+/**
+ * TX handler
+ */
+static void handle_uart_tx(struct uart_periph *p)
+{
+  // check if more data to send
+  struct SerialInit *init_struct = (struct SerialInit*)(p->init_struct);
+  chSemWait (init_struct->tx_sem);
+  while (p->tx_insert_idx != p->tx_extract_idx) {
+    uint8_t data = p->tx_buf[p->tx_extract_idx];
+    p->tx_running = true;
+    sdWrite((SerialDriver *)p->reg_addr, &data, sizeof(data));
+    p->tx_running = false;
+    // TODO send by block (be careful with circular buffer)
```

Comment:
Would it make sense in this case to internally use ChibiOS UART drivers instead of Serial drivers? Just a thought, but because UART driver uses DMA directly it is much more efficient than Serial driver. It is more complex to set up, but maybe the performance gains would be worth it - instead of sending blocks of bytes.